File: 495.teemo-attacking.cpp
/*
 * @lc app=leetcode id=495 lang=cpp
 *
 * [495] Teemo Attacking
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
    int findPoisonedDuration(vector<int>& timeSeries, int duration) {
    int size = timeSeries.size();
    int ret = 0;
    for (int i = 0; i < size - 1; i++) {
      ret += std::min(timeSeries[i + 1] - timeSeries[i], duration);
    }
    if (size > 0) {
      ret += duration;
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> timeSeries;
  timeSeries = {0};
  ASSERT_EQ(s.findPoisonedDuration(timeSeries, 0), 0);
  timeSeries = {1, 4};
  ASSERT_EQ(s.findPoisonedDuration(timeSeries, 2), 4);
  timeSeries = {1, 2};
  ASSERT_EQ(s.findPoisonedDuration(timeSeries, 2), 3);
  timeSeries = {1, 1, 4};
  ASSERT_EQ(s.findPoisonedDuration(timeSeries, 2), 4);
  return 0;
}

File: 495.teemo-attacking.rs
/*
 * @lc app=leetcode id=495 lang=rust
 *
 * [495] Teemo Attacking
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn find_poisoned_duration(time_series: Vec<i32>, duration: i32) -> i32 {
        let size = time_series.len();
        let mut ret = 0;
        for i in 0..(size - 1) {
            ret += std::cmp::min(time_series[i + 1] - time_series[i], duration);
        }
        if size > 0 {
            ret += duration;
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::find_poisoned_duration(vec![1, 4], 2), 4);
}

#[test]
fn test2() {
    assert_eq!(Solution::find_poisoned_duration(vec![1, 2], 2), 3);
}
